core/templates/components/code-mirror/codemirror-mergeview.component.ts

```diff
@@ -27,14 +28,24 @@ import { downgradeComponent } from '@angular/upgrade/static';
 export class CodemirrorMergeviewComponent implements
   AfterViewInit, OnInit, OnChanges {
   @Input() options = {};
-  @Input() leftValue;
-  @Input() rightValue;
-  codeMirrorInstance: CodeMirror.MergeView.MergeViewEditor;
-  constructor(private elementRef: ElementRef, private ngZone: NgZone) { }
+  // Below properties have been assigned as empty string if undefined.
+  // See function 'ngAfterViewInit' for reference.
+  @Input() leftValue: string | undefined;
+  @Input() rightValue: string | undefined;
+  // 'ngOnChanges' sometimes runs before the view has been initialized,
+  // to cater this we are checking it be to not undefined.
+  codeMirrorInstance!: CodeMirror.MergeView.MergeViewEditor;
+
+  constructor(
+    private elementRef: ElementRef,
+    private ngZone: NgZone,
+    private windowRef: WindowRef) { }
 
   ngOnInit(): void {
     // Require CodeMirror.
-    if (window.CodeMirror === undefined) {
+    if (
+      (this.windowRef.nativeWindow as typeof window).CodeMirror === undefined
+    ) {
```

Comment:
Shouldn't line 57 also be this.windowRef.nativeWindow?

Reply:
Done but still can't refractor the `<unknown>` pattern. We need to return a dict after spyOn
I tried this
1.
```
spyOnProperty(windowRef, 'nativeWindow`).and.returnValue(mockCodeMirror)
```
this gives `cannot read property MergeView of Undefined`
2.
```
    const mockMergeView = (element: HTMLElement): void => {
      return {
                editor: () => {
                  return { setValue: editSetValueSpy };
                },
                rightOriginal: () => {
                  return { setValue: rightOrgSetValueSpy };
                }
           }
    };
    const mockCodeMirror: typeof CodeMirror = {
      MergeView: mockMergeView
    } as typeof CodeMirror; // gives error here
    window.CodeMirror = mockCodeMirror;
    component.ngAfterViewInit();
```
3. Simply removing `<unknown>` in current changes also doesn't helps

Please suggest code changes!

Reply:
![Screenshot 2021-06-11 at 10 06 12 AM](https://user-images.githubusercontent.com/23002114/121631400-a95bb280-ca9c-11eb-9a81-675666d0c4fc.png)

The above was one example that works. You came very close to solving this with a combination of 1 and 2. They had two problems:

1. In `1.` you were not returning a 'window' like object. To be more clear, the return value of nativeWindow is basically window. So, the `returnValue.CodeMirror1 should be mockCodeMirror. So something like: ```spyOnProperty(windowRef, 'nativeWindow`).and.returnValue({CodeMirror: mockCodeMirror});```
2. In `2.`, mockCodeMIrror is not `typeof CodeMirror`. It's just an Object with the field mergeView. If you put it as typeof CodeMirror, you will get type mismatch errors. You can just spyOn nativeWindow and returnValue ```{CodeMirror: mockCodeMirror}```.

Reply:
Got it!! Thanks,will keep this in mind!